httpbp/client_middlewares.go

```diff
@@ -0,0 +1,239 @@
+package httpbp
+
+import (
+	"net/http"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/opentracing/opentracing-go"
+	"github.com/reddit/baseplate.go/breakerbp"
+	"github.com/reddit/baseplate.go/tracing"
+	"github.com/sony/gobreaker"
+)
+
+// ClientMiddleware is used to build HTTP client middlewares by implementing
+// http.RoundTripper which http.Client accepts as Transport.
+type ClientMiddleware func(roundTripper http.RoundTripper) http.RoundTripper
+
+// NewClient returns a standard HTTP client wrapped with the default
+// middlewares plus any additional client middlewares passed into this
+// function.
+func NewClient(config Config, middlewares ...ClientMiddleware) *http.Client {
+	transport := &http.Transport{
+		MaxConnsPerHost: config.MaxConnections,
+	}
+	defaults := []ClientMiddleware{
+		MonitorClient(config),
+		CircuitBreaker(config.CircuitBreaker),
+		Retries(config),
+	}
+	middlewares = append(middlewares, defaults...)
+
+	return &http.Client{
+		Transport: applyMiddlewares(transport, middlewares...),
+	}
+}
+
+func applyMiddlewares(transport http.RoundTripper, middlewares ...ClientMiddleware) http.RoundTripper {
+	// add middlewares in reverse so the first in the list is the outermost
+	for i := len(middlewares) - 1; i >= 0; i-- {
+		transport = middlewares[i](transport)
+	}
+	return transport
+}
+
+type circuitBreaker struct {
+	next     http.RoundTripper
+	settings gobreaker.Settings
+
+	mu       sync.RWMutex
+	breakers map[string]breakerbp.CircuitBreaker
+}
```

Comment:
🔕 I feel that those structs could be avoided if you just define a lambda type:

```go
type roundtripper func(req *http.Request) (*http.Response, error)

func (f roundtripper) RoundTrip(req *http.Request) (*http.Response, error) {
  return f(req)
}
```

this way all those structs can be actually inlined.

Reply:
:+1: that would also be in-line with our other middleware implementations.

Reply:
The only caveat is that it cannot be passed directly to `http.Client` as transport as it expects `http.RoundTripper` but I think this can be covered by having an abstraction that will take care of delegating it to the `roundTripper` type.

Reply:
Currently trying to apply this approach: would this require a `func type` definition for each middleware?

Reply:
I refactored the client middleware interface, let me know if this is what you had in mind or if there is another way to reduce the complexity further.

Reply:
@fishy I updated it based on your Go playground snippet. This is a lot shorter now.